File: internal/engine/movie_frames.go
package engine

import (
	"container/list"
	"slices"

	"github.com/itskenny0/titanic-godot/internal/df"
)

type MovieImage struct {
	Pixels        []byte
	Width, Height int
	ID            uint64
}
type movieImageEntry struct {
	index int
	image *MovieImage
}
type movieCheckpoint struct {
	index  int
	buffer df.FrameBuffer
	shown  []byte
}

func (c *movieCheckpoint) bytes() int {
	return len(c.buffer.Pixels) + len(c.buffer.ZPixels) + len(c.shown)
}

// MovieFrames decodes deltas on demand. Sparse restart points bound seek cost
// without retaining the entire credits or intro movie in handheld memory.
type MovieFrames struct {
	Segment                       *df.MovieSegment
	ImageBudget, CheckpointBudget int
	TransparentV1                 bool
	DecodedFrames                 int
	fb                            df.FrameBuffer
	position                      int
	shown                         []byte
	images, checkpoints           list.List
	imageIndex, checkpointIndex   map[int]*list.Element
	imageBytes, checkpointBytes   int
}

func NewMovieFrames(seg *df.MovieSegment) *MovieFrames {
	return &MovieFrames{Segment: seg, ImageBudget: 2 << 20, CheckpointBudget: 4 << 20, position: -1, imageIndex: map[int]*list.Element{}, checkpointIndex: map[int]*list.Element{}}
}
func (m *MovieFrames) Len() int { return len(m.Segment.Frames) }
func (m *MovieFrames) RetainedBytes() int {
	return m.imageBytes + m.checkpointBytes + len(m.fb.Pixels) + len(m.fb.ZPixels) + len(m.shown)
}
func (m *MovieFrames) Get(index int) (*MovieImage, error) {
	if index < 0 || index >= m.Len() {
		return nil, nil
	}
	if e := m.imageIndex[index]; e != nil {
		m.images.MoveToBack(e)
		return e.Value.(movieImageEntry).image, nil
	}
	if index <= m.position {
		m.restore(index)
	}
	var result *MovieImage
	for m.position < index {
		next := m.position + 1
		meta := m.Segment.Frames[next]
		f, err := df.DecodeFrame(m.Segment.File.Data(meta.LocationFrame), &m.fb, m.Segment.File.Order)
		if err != nil {
			return nil, err
		}
		m.DecodedFrames++
		if next == index || m.TransparentV1 {
			pixels := slices.Clone(m.fb.Pixels[:f.Width*f.Height])
			if m.TransparentV1 {
				if len(m.shown) == len(pixels) {
					for i, p := range pixels {
						if p == 0 || p == 255 {
							pixels[i] = m.shown[i]
						}
					}
				}
				m.shown = pixels
			}
			if next == index {
				result = &MovieImage{Pixels: pixels, Width: f.Width, Height: f.Height, ID: newDrawID()}
			}
		}
		m.position = next
		if next%32 == 0 {
			m.checkpoint()
		}
	}
	if result != nil {
		m.imageIndex[index] = m.images.PushBack(movieImageEntry{index, result})
		m.imageBytes += len(result.Pixels)
		for m.imageBytes > m.ImageBudget && m.images.Len() > 1 {
			e := m.images.Front()
			old := e.Value.(movieImageEntry)
			m.imageBytes -= len(old.image.Pixels)
			delete(m.imageIndex, old.index)
			m.images.Remove(e)
		}
	}
	return result, nil
}
func (m *MovieFrames) checkpoint() {
	if m.checkpointIndex[m.position] != nil {
		return
	}
	size := len(m.fb.Pixels) + len(m.fb.ZPixels) + len(m.shown)
	if size > m.CheckpointBudget {
		return
	}
	c := movieCheckpoint{index: m.position, buffer: df.FrameBuffer{Pixels: slices.Clone(m.fb.Pixels), ZPixels: slices.Clone(m.fb.ZPixels), Width: m.fb.Width, Height: m.fb.Height}, shown: slices.Clone(m.shown)}
	m.checkpointIndex[m.position] = m.checkpoints.PushBack(c)
	m.checkpointBytes += size
	for m.checkpointBytes > m.CheckpointBudget {
		e := m.checkpoints.Front()
		old := e.Value.(movieCheckpoint)
		m.checkpointBytes -= old.bytes()
		delete(m.checkpointIndex, old.index)
		m.checkpoints.Remove(e)
	}
}
func (m *MovieFrames) restore(index int) {
	nearest := -1
	var selected *list.Element
	for at, e := range m.checkpointIndex {
		if at < index && at > nearest {
			nearest, selected = at, e
		}
	}
	m.fb = df.FrameBuffer{}
	m.shown = nil
	m.position = nearest
	if selected != nil {
		c := selected.Value.(movieCheckpoint)
		m.fb = df.FrameBuffer{Pixels: slices.Clone(c.buffer.Pixels), ZPixels: slices.Clone(c.buffer.ZPixels), Width: c.buffer.Width, Height: c.buffer.Height}
		m.shown = slices.Clone(c.shown)
		m.checkpoints.MoveToBack(selected)
	}
}

File: engine/movie-frames.ts
import {FrameBuffer,decodeFrame} from '@dreamfactory/engine/df/image';
import {compositeFrameV1} from '@dreamfactory/engine/df/mov-v1';
import type {MovSegment} from '@dreamfactory/engine/df/mov';

export interface MovieImage {pixels:Uint8Array;width:number;height:number;}
interface Checkpoint {pixels:Uint8Array;z:Uint8Array;width:number;height:number;shown:Uint8Array|null;}

// Decoding the supplied credits segment used about 158 MiB before it could start.
// Keep a few recent images and sparse restart points instead. The raw delta
// buffer is independent of presented images, so later frames cannot alter one
// still in use by the renderer.
export class MovieFrames {
 readonly length:number;
 private fb=new FrameBuffer();
 private position=-1;
 private shown:Uint8Array|null=null;
 private images=new Map<number,MovieImage>();
 private checkpoints=new Map<number,Checkpoint>();
 private imageBytes=0;
 private checkpointBytes=0;
 decodedFrames=0;
 constructor(private seg:MovSegment,private imageBudget=2*1024*1024,private checkpointBudget=4*1024*1024) {
  this.length=seg.frames.length;
 }
 get retainedBytes(){return this.imageBytes+this.checkpointBytes+this.fb.pixels.byteLength+this.fb.zPixels.byteLength+(this.shown?.byteLength||0);}
 get(index:number):MovieImage|undefined {
  if(!Number.isInteger(index)||index<0||index>=this.length)return undefined;
  const cached=this.images.get(index);
  if(cached){this.images.delete(index);this.images.set(index,cached);return cached;}
  if(index<=this.position)this.restore(index);
  let result:MovieImage|undefined;
  while(this.position<index) {
   const next=this.position+1,meta=this.seg.frames[next];
   const d=decodeFrame(this.seg.file.containers[meta.locationFrame].data,this.fb,this.seg.file.order);
   this.decodedFrames++;
   // Intermediate images are not needed unless v1 transparency composites them.
   if(next===index||this.seg.dfV1) {
    const pixels=this.fb.pixels.slice(0,d.width*d.height);
    if(this.seg.dfV1){compositeFrameV1(pixels,this.shown);this.shown=pixels;}
    if(next===index)result={pixels,width:d.width,height:d.height};
   }
   this.position=next;
   if(next%32===0)this.checkpoint();
  }
  if(result){this.images.set(index,result);this.imageBytes+=result.pixels.byteLength;this.trimImages();}
  return result;
 }
 private trimImages(){
  while(this.imageBytes>this.imageBudget&&this.images.size>1){const key=this.images.keys().next().value!;this.imageBytes-=this.images.get(key)!.pixels.byteLength;this.images.delete(key);}
 }
 private checkpoint(){
  if(this.checkpoints.has(this.position))return;
  const c={pixels:this.fb.pixels.slice(),z:this.fb.zPixels.slice(),width:this.fb.width,height:this.fb.height,shown:this.shown?.slice()??null};
  const bytes=c.pixels.byteLength+c.z.byteLength+(c.shown?.byteLength||0);
  if(bytes>this.checkpointBudget)return;
  this.checkpoints.set(this.position,c);this.checkpointBytes+=bytes;
  while(this.checkpointBytes>this.checkpointBudget){
   const key=this.checkpoints.keys().next().value!,old=this.checkpoints.get(key)!;
   this.checkpointBytes-=old.pixels.byteLength+old.z.byteLength+(old.shown?.byteLength||0);this.checkpoints.delete(key);
  }
 }
 private restore(index:number){
  let nearest=-1;
  for(const at of this.checkpoints.keys())if(at<index&&at>nearest)nearest=at;
  this.fb=new FrameBuffer();this.shown=null;this.position=nearest;
  if(nearest>=0){
   const c=this.checkpoints.get(nearest)!;
   this.fb.pixels=c.pixels.slice();this.fb.zPixels=c.z.slice();this.fb.width=c.width;this.fb.height=c.height;this.shown=c.shown?.slice()??null;
   this.checkpoints.delete(nearest);this.checkpoints.set(nearest,c);
  }
 }
}
